woocommerce/woocommerce-android
# Refactor Review Moderation logic to decouple ReviewListFragment and ReviewDetailFragment

Part of: #5647

⚠️ Please don't merge, the FluxC PR needs to be merged first: https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2316

Only one reviewer is needed, but feel free to review if you have time 🙏  

### Description
The review moderation currently is causing a tight couple between `ReviewListFragment` and `ReviewDetailFragment`, which prevents reusing the same logic in `ProductReviewsFragment`. The code also is hard to follow, it uses EventBus notifications, and the logic is spread across ViewModel and UI (Fragment and Adapter).
A temporary solution was to hide the moderation buttons when viewing the opening `ReviewDetailFragment` from `ProductReviewsFragment` (check https://github.com/woocommerce/woocommerce-android/pull/3003).

This is a first PR in the refactoring, it brings just the same functionality as before, which means successive moderation requests are not handled well (before the moderation request was lost if a new one was submitted before the first one is submitted, while now there is just a UI issue where the shown status will be incorrect during a short time).
The functionality will be improved in a second PR, and a third PR will bring moderation to `ProductReviewsFragment`.

### Architecture
The bulk of the logic is part of the `ReviewModerationHandler` class which has the following API:

```kotlin
class ReviewModerationHandler {
   val pendingModerationStatus: Flow<ReviewModerationStatus>
   fun postModerationRequest(review: ProductReview, newStatus: ProductReviewStatus)
   fun undoLastOperation()
}
```
And to help reuse code for ViewModels and Fragments, and enforce a contract, I added two interfaces:
1. `ReviewModerationConsumer` to be implemented by ViewModels who want to handle moderation events.
2. `ReviewModerationUi` to be implemented by corresponding Fragments.
If you can think of any improvement to the architecture, please let me know 🙏

### Testing instructions

1. Create some reviews on your store.
3. Open the Reviews list screen on the app.
4. Tap on one of the reviews.
5. Click on one of the moderation options (SPAM and TRASH should trigger a removal of the review from the list).
6. Wait until the snackbar is hidden.
7. Confirm the review status has been correctly updated.
8. Repeat step 4.
9. Click on "Undo" in the snackbar.
10. Confirm the action is canceled.
